### PR TITLE
DualTor data plane IO test utilities

### PR DESCRIPTION
**Pull request by vaibhavhd**
### Description of PR

Summary: Data plane I/O utilities and fixtures for dual-TOR tests
Fixes: https://github.com/Azure/sonic-mgmt/issues/2759

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Approach
Utilities for dualTOR data plane testing. At high level, this PR provides two utilities for the Dual tor tests to consume:
1. `send_t1_to_server_after_action` - simulate data plane traffic from T1 router to server
2. `send_server_to_t1_after_action` - simulate data plane traffic from server to T1 router

#### How did you do it?
The utilities are provided as fixtures which yield the data-plane-io functionality to be used within tests.
The I/O test runs mostly on sonic-mgmt-container utilizing the PTF-adapter utility, except the sniffer part, which runs on the PTF container due to permission issues and socket ports that scappy sniffer demands.

Two types of I/O support is provided in two directions - T1 to server, and server to T1.

#### How did you verify/test it?
Tested on a single TOR setup, and the traffic disruptions are successfully captured. Also tried with non-disruptive test, and no disruption and packet loss was seen. Sample tests used are below:

```Python
def try_config_reload(duthost):
    config_reload = duthost.shell("config reload -y")['stdout']
    logger.info("Output of action: {}".format(config_reload))


def try_bgp_summary(duthost):
    show_bgp_summary = duthost.shell("show ip bgp summary")['stdout']
    logger.info("Output of action: {}".format(show_bgp_summary))


def test_t1_to_server_active_tor_config_reload(duthosts, rand_one_dut_hostname, send_t1_to_server_after_action):
    duthost = duthosts[rand_one_dut_hostname]
    
    logger.info("Active ToR before {}".format(duthost.hostname))

    tor_port = 29
    send_t1_to_server_after_action(duthost,
                                tor_port=tor_port,
                                action = lambda: try_config_reload(duthost))
    # TODO add more verification steps here

    logger.info("Active ToR after {}".format(duthost.hostname))

def test_server_to_t1_active_tor_bgp_summary(duthosts, rand_one_dut_hostname, send_server_to_t1_after_action):
    duthost = duthosts[rand_one_dut_hostname]
    #active_host_before = get_active_torhost()
    logger.info("Active ToR before {}".format(duthost.hostname))

    # perform bgp shutdown on active TOR
    send_server_to_t1_after_action(duthost,
                                action = lambda: try_bgp_summary(duthost))
    # TODO add more verification steps here

    logger.info("Active ToR after {}".format(duthost.hostname))
```

### CODE REVIEW

File: tests/common/dualtor/dual_tor_io.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+        self.send_interval = 0.0035 # Inter-packet interval
+        self.packets_to_send = min(int(self.time_to_listen / (self.send_interval + 0.0015)), 45000) # How many packets to be sent in send_in_background method
+
+        self.dataplane = self.ptfadapter.dataplane
+        self.dataplane.flush()
+        self.total_disrupt_time = None
+        self.disrupts_count = None
+        self.total_disrupt_packets = None
+        # This list will contain all unique Payload ID, to filter out received floods.
+        self.unique_id = list()
```
Comment (theasianpianist) [inline tests/common/dualtor/dual_tor_io.py:50]: What's the eventual expected size of this? Would we see any meaningful performance improvements from using a `set` instead?

File: tests/common/dualtor/dual_tor_io.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+        tcp_dport = 5000
+        tcp_tx_packet = testutils.simple_tcp_packet(
+                      eth_dst=self.dut_mac,
+                      ip_src=self.from_server_src_addr,
+                      ip_dst=self.from_server_dst_addr,
+                      tcp_dport=tcp_dport
+                 )
+
+        self.packets_list = []
+        for i in xrange(self.packets_to_send):
```
Comment (theasianpianist) [inline tests/common/dualtor/dual_tor_io.py:171]: Prefer `range` for Python3 compatibility?

File: tests/common/dualtor/dual_tor_io.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+                    sent_payload = int(str(packet[scapyall.TCP].payload).replace('X',''))
+                    sent_packets[sent_payload] = packet.time
+                    continue
+                if packet[scapyall.Ether].src == dut_mac:
+                    # This is a received packet.
+                    received_time = packet.time
+                    received_payload = int(str(packet[scapyall.TCP].payload).replace('X',''))
+                    received_counter += 1
+                if not (received_payload and received_time):
+                    # This is the first valid received packet.
```
Comment (theasianpianist) [inline tests/common/dualtor/dual_tor_io.py:343]: Should we verify that the first packet we receive has payload = 1?

File: tests/common/dualtor/dual_tor_io.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+                        disruption_start = datetime.datetime.fromtimestamp(prev_time)
+                    disruption_stop = datetime.datetime.fromtimestamp(received_time)
+                prev_payload = received_payload
+                prev_time = received_time
+            if received_counter == 0:
+                logger.error("Sniffer failed to filter any traffic from DUT")
+            else:
+                logger.info("Total number of filtered incoming packets captured {}".format(received_counter))
+            if lost_packets:
+                logger.info("Disruptions happen between %s and %s." % \
```
Comment (theasianpianist) [inline tests/common/dualtor/dual_tor_io.py:364]: Looks like we are just logging the last disruption to happen, any way to capture all of them?

**Review by theasianpianist — CHANGES_REQUESTED** [review]

File: tests/common/dualtor/data_plane_utils.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    Args:
+        ptfhost (fixture): Fixture for PTF instance to be used during the test
+        ptfadapter (fixture): Ficture which provides helper utility to use ptf ptf testutils
+        tbinfo (fixture): Fixture for testebd inventory  information
+
+    Yields:
+        function: A helper function to run and monitor the IO test
+    """
+    logger.info('Copy ARP responder to the PTF container  {}'.format(ptfhost.hostname))
+    ptfhost.copy(src='scripts/arp_responder.py', dest='/opt')
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py]: Yes, but we are now doing a little more than that - supervisorctl control, arp args and conf. I rather moved this set of logic (which is common to two fixtures here) to a helper function in this file.

File: tests/common/dualtor/data_plane_utils.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    logging.info("Enabling arp_responder")
+    ptfhost.shell("supervisorctl reread && supervisorctl update")
+    ptfhost.shell("supervisorctl restart arp_responder")
+    logging.info("arp_responder enabled")
+
+    duthosts = []
+    def t1_to_server_io_test(duthost, server_port=None, tor_port=None, expect_tunnel_packet=False, delay=1, timeout=5, action=None):
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py]: We don't need to pass args separately. `lambda` does the trick here binding the method with args. Like the one that is shown in example test:
`action = lambda: try_config_reload(duthost)`

File: tests/common/dualtor/data_plane_utils.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        ptfadapter (fixture): Ficture which provides helper utility to use ptf ptf testutils
+        tbinfo (fixture): Fixture for testebd inventory  information
+
+    Yields:
+        function: A helper function to run and monitor the IO test
+    """
+    logger.info('Copy ARP responder to the PTF container  {}'.format(ptfhost.hostname))
+    ptfhost.copy(src='scripts/arp_responder.py', dest='/opt')
+    logging.info("Enabling arp_responder")
+    ptfhost.shell("supervisorctl reread && supervisorctl update")
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py]: You are right. Added ARP responder config.

File: tests/common/dualtor/data_plane_utils.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        duthosts.append(duthost)
+        io_ready = threading.Event()
+        tor_IO = DualTorIO(duthost, ptfhost, ptfadapter, tbinfo, server_port, tor_port, expect_tunnel_packet, delay, timeout, io_ready)
+        send_and_sniff = threading.Thread(target=tor_IO.start_io_test, kwargs={'traffic_generator': tor_IO.generate_from_t1_to_server})
+        send_and_sniff.start()
+        if action:
+            # do not perform the provided action until IO threads (sender and sniffer) are ready
+            io_ready.wait()
+            logger.info("Sender and sniffer threads started, ready to execute the callback action")
+            action()
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py:66]: Replied above.

File: tests/common/dualtor/data_plane_utils.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        logger.info('Clearing arp entries on DUT  {}'.format(duthost.hostname))
+        duthost.shell('sonic-clear arp')
+        logger.info('Clearing all fdb entries on DUT  {}'.format(duthost.hostname))
+        duthost.shell('sonic-clear fdb all')
+
+
+@pytest.fixture
+def send_server_to_t1_after_action(ptfhost, ptfadapter, tbinfo):
+    """
+    Starts IO test from T1 router to server.
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py]: Done

File: tests/common/dualtor/data_plane_utils.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def send_t1_to_server_after_action(ptfhost, ptfadapter, tbinfo):
+    """
+    Starts IO test from T1 router to server.
+    As part of IO test the banckground thread sends and sniffs packets.
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py]: Done

File: tests/common/dualtor/data_plane_utils.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+    duthosts = []
+    def t1_to_server_io_test(duthost, server_port=None, tor_port=None, expect_tunnel_packet=False, delay=1, timeout=5, action=None):
+        """
+        Helper method for `send_t1_to_server_after_action`.
+        Starts sender and sniffer before performing the action on the tor host.
+
+        Args:
+            server_port: The port intended to receive the packet
+            tor_port: The port through which to send the packet. Connected to either the upper or lower ToR.
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py]: Done

File: tests/common/dualtor/data_plane_utils.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    As part of IO test the banckground thread sends and sniffs packets.
+    As soon as sender and sniffer threads are in running state, a callback action is performed.
+    When action is finished, the sender and sniffer threads are given time to complete.
+    Finally, the collected packets are sniffed, and the disruptions are measured.
+
+    Should return `send_t1_to_server` for the external user to control the timing of test.
+    As part of teardown, the ARP, FDB tables are cleared and ptf dataplane is flushed.
+    Args:
+        ptfhost (fixture): Fixture for PTF instance to be used during the test
+        ptfadapter (fixture): Ficture which provides helper utility to use ptf ptf testutils
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py]: Done

File: tests/common/dualtor/data_plane_utils.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        tbinfo (fixture): Fixture for testebd inventory  information
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/data_plane_utils.py]: Done

File: tests/common/dualtor/dual_tor_io.py [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+        # - "Hosts" (IP/MAC pairs) are distributed evenly amongst the ports in the VLAN
+        addr_list = list(IPNetwork(self.vlan_network))
+        for counter, i in enumerate(xrange(2, VLAN_HOSTS + 2)):
+            mac = VLAN_BASE_MAC_PATTERN.format(counter)
+            port = self.vlan_ports[i % len(self.vlan_ports)]
+            addr = random.choice(addr_list)
+            # Ensure that we won't get a duplicate ip address
+            addr_list.remove(addr)
+
+            vlan_host_map[port][str(addr)] = mac
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/dual_tor_io.py:89]: Yes. MAC to IP address mapping for vlan ports.

File: tests/common/dualtor/dual_tor_io.py (hunk 1 above)
Comment (vaibhavhd) [reply tests/common/dualtor/dual_tor_io.py:51]: In the tests performed so far, this ranges between 50k to 70k.

File: tests/common/dualtor/dual_tor_io.py (hunk 2 above)
Comment (vaibhavhd) [reply tests/common/dualtor/dual_tor_io.py]: Done

File: tests/common/dualtor/dual_tor_io.py (hunk 3 above)
Comment (vaibhavhd) [reply tests/common/dualtor/dual_tor_io.py:356]: The first packet actually has payload 0 and timestamp as 0 - which is being checked here. May be more explicit check is better?

File: tests/common/dualtor/dual_tor_io.py (hunk 4 above)
Comment (vaibhavhd) [reply tests/common/dualtor/dual_tor_io.py]: All disruptions (between consecutive packets) are already being captured inside the packet loop above. This log is the high-level loss to understand if the start and end range corresponds to the timestamp between which action is performed.

File: tests/common/dualtor/dual_tor_io.py (hunk 1 above)
Comment (theasianpianist) [reply tests/common/dualtor/dual_tor_io.py:52]: Have you seen any performance issues?

File: tests/common/dualtor/dual_tor_io.py (hunk 3 above)
Comment (theasianpianist) [reply tests/common/dualtor/dual_tor_io.py:374]: Misunderstood the first time, looks good to me

File: tests/common/dualtor/dual_tor_io.py (hunk 1 above)
Comment (vaibhavhd) [reply tests/common/dualtor/dual_tor_io.py]: Nothing measurable in seconds for this range. But, this is a useful change, should we ever want the test to run for longer duration. The `in` check with set is significantly faster.
Added a new commit to change the type from `list` to `set`.

**Review by theasianpianist — APPROVED** [review]

**Comment by vaibhavhd** [comment]
Not merging this PR yet as the verification is blocked, due to some issues in simulating switchover scenario and the lack of ICMP responder in PTF to relay link manager packets.

**Comment by vaibhavhd** [comment]
This is ready for merge now. The code has been tested with the switchover scenario, finally. Upon simulating a switchover, the traffic was disrupted for 0.071s (<1s), and 19 packets were lost.
These numbers are within the present expected range for the ToR switchover.

```
09:53:48 INFO dual_tor_io.py:examine_each_packet:404: Total number of filtered incoming packets captured 35981
09:53:48 INFO dual_tor_io.py:examine_each_packet:406: Disruptions happen between 2021-02-13 09:41:58.938589 and 2021-02-13 09:41:58.835691.
09:53:48 INFO dual_tor_io.py:examine_flow:466: Packet flow examine finished after 0:01:44.829916
09:53:48 ERROR dual_tor_io.py:start_io_test:140: The longest disruption lasted 0.008 seconds. 2 packet(s) lost.
09:53:48 ERROR dual_tor_io.py:start_io_test:142: Total disruptions count is 16. All disruptions lasted 0.071 seconds. Total 19 packet(s) lost
FAILED             
```

**Pull request merged**